src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.IncrementalAnalyzerProcessor.cs

```diff
@@ -100,11 +105,73 @@ public void Enqueue(WorkItem item)
                 {
                     Contract.ThrowIfNull(item.DocumentId);
 
-                    _highPriorityProcessor.Enqueue(item);
-                    _normalPriorityProcessor.Enqueue(item);
-                    _lowPriorityProcessor.Enqueue(item);
+                    var options = _registration.Workspace.Options;
+                    var analysisScope = SolutionCrawlerOptions.GetBackgroundAnalysisScope(options);
+
+                    if (ShouldEnqueueForAllQueues(item, analysisScope))
+                    {
+                        _highPriorityProcessor.Enqueue(item);
+                        _normalPriorityProcessor.Enqueue(item);
+                        _lowPriorityProcessor.Enqueue(item);
+                    }
+                    else
+                    {
+                        if (ShouldEnqueueReducedWorkItem(item, _highPriorityProcessor.Analyzers, options, analysisScope, out var reducedWorkItem))
+                        {
+                            _highPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
+
+                        if (ShouldEnqueueReducedWorkItem(item, _normalPriorityProcessor.Analyzers, options, analysisScope, out reducedWorkItem))
+                        {
+                            _normalPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
+
+                        if (ShouldEnqueueReducedWorkItem(item, _lowPriorityProcessor.Analyzers, options, analysisScope, out reducedWorkItem))
+                        {
+                            _lowPriorityProcessor.Enqueue(reducedWorkItem.Value);
+                        }
```

Comment:
shoudl these be else-ifs?  couldn't this end up with the same item added to multiple queues?  is that what we want?  why?

Reply:
I am retaining the original semantics where a work item is always enqueued in all the queues. For the else part here we enqueue based on whether it has any applicable analyzers in the queue. I plan to do follow-up investigation on the point you have raised - I don't know if we need to enqueue the work item on all queues, and if that is redundant we should change these into else-ifs.

Reply:
I investigated further and found out that each queue has different semantics and further checks in its `Enqueue` call. For example, high priority queue only processes work items which have `InvocationReasons.SyntaxChanged` as one of the invocation reason. This ensures we prioritize syntax analysis and diagnostics over semantic analysis, the latter being performed by the normal priority queue. Low priority queue only processes project level events.

Reply:
that is super confusing :)   why not call the high-pri queue the SyntaxChangedQueue or something? or, if it is going to filter, have the filtering be done here.  having to know that you queue up everything to every queue, and then the queues actually throw some out is just a confusing design IMO.

Reply:
Sure, I will refactor this a bit more in a later PR.